### ISSUE TITLE
input and torch.tensor or tensor?

### ISSUE THREAD

**Issue by muqishan** [triton-inference-server/server#5735]
When there are times when some preprocessing must be placed locally, but the data volume is relatively large, I choose to use torch to complete the preprocessing on the GPU, such as batch_ imgs.to('cuda')。 This can speed up my preprocessing by a hundred times. But when it comes to reasoning
cudashm.set_ shared_ memory_ region(shm_input_handle, [batch_imgs])
batch_ IMGS requires batch_ Imgs. cpu(). numpy(), when the data volume is large, this is a time-consuming behavior, such as on RTX3090TI, batch_ The shape of IMGS is [64 * 3 * 360 * 640], which will take 80ms. Can we consider directly receiving torch. sensor or tensor?

**Comment by GuanLuo** [comment]
Let me rephrase your use case to make sure I understand it correctly:

You have some preprocessed data that is already on GPU on the client side, which then you want to pass to server via CUDA shared memory. But due to the limitation on Triton's `cudashm.set_ shared_ memory_ region` API, it only accepts data on system memory and causes extra data copy.

If that is the case, I think it is a reasonable ask to enhance the `cudashm` to accept data on GPU. @Tabrizian what do you think?

**Comment by muqishan** [comment]
Yes, thank you for your response. I have used CUDA shared memory to send data to Triton, but before creating the CUDA shared memory and moving the data to be inferred to the shared memory, I need to use a framework like PyTorch, TensorFlow, or any other framework with tensors to preprocess the data, such as torch.to('cuda'), so that I can complete the preprocessing on the GPU. However, Triton's API can only accept NumPy arrays. Therefore, before placing the data into shared memory, I have to move the data back to the CPU, such as with data.cpu().numpy(). Why not consider having Triton directly accept tensors (from Torch or TensorFlow) and place them into shared memory?

**Comment by muqishan** [comment]
Alternatively, consider a different scenario: due to the limited concurrent capabilities of the CPU, I plan to use the GPU to decode multiple camera feeds. As a result, the frames are already on the GPU device. However, due to API limitations, even though inference can be completed on the GPU using CUDA shared memory, the data still needs to be moved to the CPU and then back to the GPU, resulting in two additional data copy operations. On the NVIDIA A10, it takes 80ms to move a 64x3x360x640 matrix from the GPU to the CPU. Although this is already quite fast, it is a critical overhead in industrial applications.

**Comment by muqishan** [comment]
Perhaps Triton offers a better solution for this scenario, but as a deployment beginner, it is difficult for me to find clues for a better solution in the client documentation. Please understand the confusion of a deployment novice.
